bluemira/power_cycle/base.py

```diff
@@ -0,0 +1,468 @@
+"""
+Base classes for the power cycle model.
+"""
+
+# Import
+import abc
+
+import matplotlib.pyplot as plt
+import numpy as np
+
+# from typing import Union
+
+
+# ######################################################################
+# CLASS PROPERTY DECORATOR
+# ######################################################################
+class classproperty(object):
+    def __init__(self, fget):
+        self.fget = fget
+
+    def __get__(self, owner_self, owner_cls):
+        return self.fget(owner_cls)
+
+
+# ######################################################################
+# POWER CYCLE ABSTRACT BASE CLASS
+# ######################################################################
+
+
+class PowerCycleABC(abc.ABC):
+    """
+    Abstract base class for classes in the Power Cycle module.
+
+    Parameters
+    ----------
+    name: `str`
+        Description of the instance.
+    """
+
+    # ------------------------------------------------------------------
+    # CLASS ATTRIBUTES
+    # ------------------------------------------------------------------
+
+    # Plot defaults (arguments for `matplotlib.pyplot.plot`)
+    _plot_kwargs = {
+        "c": "k",  # Line color
+        "lw": 2,  # Line width
+        "ls": "-",  # Line style
+    }
+    _scatter_kwargs = {
+        "c": "k",  # Marker color
+        "s": 100,  # Marker size
+        "marker": "x",  # Marker style
+    }
+
+    # Plot text settings (for `matplotlib.pyplot.text`)
+    _text_angle = 45  # rotation angle
+    _text_index = -1  # index of (time,data) point used for location
+
+    # ------------------------------------------------------------------
+    # CONSTRUCTOR
+    # ------------------------------------------------------------------
+
+    def __init__(self, name: str):
+
+        # Store name
+        self.name = self._validate_name(name)
+
+        # Dynamically add validation classmethod
+        # setattr(a, 'func', classmethod(func))
+
+    @classmethod
+    def _validate_name(cls, name):
```

Comment:
This is not a classmethod you can get what you want from `self.__class__.__name__`.

More generally this seems like overkill on the validation it seems odd that you are separating the errors from the location they need to be raised. I would remove this

Reply:
point 1 as we talked about is something we possibly can deal with later and for now just duplicate.

For point 2 it is then not a `staticmethod` because it requires information from the instance of the class (what methods exist). You should nearly always have to initialise classes as a `classmethod` should modify the class otherwise there is no need for it. 

Modifying a class rather than an instance of the class is dangerous because if a new instance of that class is created that will also have been modified.